rust/src/interface.rs

```diff
@@ -0,0 +1,295 @@
+    if spos % 2 == 1 || spos > SEPARATOR_POSITION {
+        // The separator must be in a valid location
+        return false;
+    }
+    if code.len() - spos - 1 == 1 {
+        // There must be > 1 character after the separator
+        return false;
+    }
+
+    // Validate padding
+    let padstart = code.find(PADDING_CHAR);
+    if padstart.is_some() {
+        let ppos = padstart.unwrap();
+        if ppos == 0 || ppos % 2 == 1 {
+            // Padding must be "within" the string, starting at an even position
+            return false;
+        }
+        if code.len() > spos + 1 {
+            // If there is padding, the code must end with the separator char
+            return false;
+        }
+        let eppos = code.rfind(PADDING_CHAR).unwrap();
+        if eppos - ppos % 2 == 1 {
+            // Must have even number of padding chars
+            return false;
+        }
+        // Extract the padding from the code (mutates code)
+        let padding: String = code.drain(ppos..eppos+1).collect();
+        if padding.chars().any(|c| c != PADDING_CHAR) {
+            // Padding must be one, contiguous block of padding chars
+            return false;
+        }
+    }
+
+    // Validate all characters are permissible
+    code.chars()
+        .map(|c| c.to_ascii_uppercase())
+        .all(|c| c == SEPARATOR || CODE_ALPHABET.contains(&c))
+}
+
+/// Determines if a code is a valid short code.
+///
+/// A short Open Location Code is a sequence created by removing four or more
+/// digits from an Open Location Code. It must include a separator character.
+pub fn is_short(_code: &str) -> bool {
+    is_valid(_code) &&
+        _code.find(SEPARATOR).unwrap() < SEPARATOR_POSITION
+}
+
+/// Determines if a code is a valid full Open Location Code.
+///
+/// Not all possible combinations of Open Location Code characters decode to
+/// valid latitude and longitude values. This checks that a code is valid
+/// and also that the latitude and longitude values are legal. If the prefix
+/// character is present, it must be the first character. If the separator
+/// character is present, it must be after four characters.
+pub fn is_full(_code: &str) -> bool {
+    is_valid(_code) && !is_short(_code)
+}
+
+/// Encode a location into an Open Location Code.
+///
+/// Produces a code of the specified length, or the default length if no
+/// length is provided.
+/// The length determines the accuracy of the code. The default length is
+/// 10 characters, returning a code of approximately 13.5x13.5 meters. Longer
+/// codes represent smaller areas, but lengths > 14 are sub-centimetre and so
+/// 11 or 12 are probably the limit of useful codes.
+pub fn encode(pt: Point<f64>, code_length: usize) -> String {
+    let mut lat = clip_latitude(pt.lat());
+    let mut lng = normalize_longitude(pt.lng());
+
+    // Latitude 90 needs to be adjusted to be just less, so the returned code
+    // can also be decoded.
+    if lat == LATITUDE_MAX {
```

Comment:
Using == to compare floats can be optimistic. Maybe use a comparison that the difference between them is very, very small, or that lat is > LATITUDE_MAX? e.g.
`if lat > LATITUDE_MAX || (LATITUDE_MAX - lat) < 0.0000000001f64 {`